### PR TITLE
img-remote: introduction of criu-image-streamer [v2]

### PR DESCRIPTION
**Pull request by nviennot**
v1 is https://github.com/checkpoint-restore/criu/pull/1087

This patch series applies on master to make it easier to backport the criu-dev branch onto master.

Changes:
* The test suite takes a `--criu-image-streamer-dir` argument
* reworded the comments a bit

### CODE REVIEW

**Comment by avagin** [comment]
LGTM

File: criu/pipes.c [hunk 1]
```diff
@@ -463,18 +463,16 @@ int dump_one_pipe_data(struct pipe_data_dump *pd, int lfd, const struct fd_parms
...
 		goto err_close;
 
-	if (bytes) {
+	while (bytes > 0) {
 		int wrote;
-
 		wrote = splice(steal_pipe[0], NULL, img_raw_fd(img), NULL, bytes, 0);
 		if (wrote < 0) {
 			pr_perror("Can't push pipe data");
 			goto err_close;
```
Comment (rppt) [inline criu/pipes.c:471]: Maybe we also should add a check for `-EAGAIN` here and continue the loop, or at least retry several more times

**Comment by rst0git** [comment]
hmm, the following test case fails with `--stream` ([restore.log](https://github.com/checkpoint-restore/criu/files/4896832/restore.log))

```sh
# python zdtm.py run --stream -f ns -t zdtm/static/socket-tcp-keepalive
criu-image-streamer 1.0.0
=== Run 1/1 ================ zdtm/static/socket-tcp-keepalive
================== Run zdtm/static/socket-tcp-keepalive in ns ==================
Start test
./socket-tcp-keepalive --pidfile=socket-tcp-keepalive.pid --outfile=socket-tcp-keepalive.out
Run criu-image-streamer in capture mode
Run criu dump
Run criu-image-streamer in extract mode
Run criu-image-streamer in serve mode
Run criu restore
=[log]=> dump/zdtm/static/socket-tcp-keepalive/51/1/restore.log
------------------------ grep Error ------------------------
b'RTNETLINK answers: File exists'
b'RTNETLINK answers: File exists'
b'(00.009915)      1: \tRunning ip route restore'
b'Failed to restore: ftell: Illegal seek'
b'(00.011302)      1: Error (criu/util.c:624): exited, status=1'
b'(00.011319)      1: Error (criu/net.c:1818): IP tool failed on route restore'
b"(00.011359)      1: Error (criu/util.c:1389): Can't wait or bad status: errno=0, status=65280"
b'(00.015290) Error (criu/cr-restore.c:2401): Restoring FAILED.'
------------------------ ERROR OVER ------------------------
########## Test zdtm/static/socket-tcp-keepalive FAIL at CRIU restore ##########
##################################### FAIL #####################################
```

**Comment by nviennot** [comment]
Regarding `b'Failed to restore: ftell: Illegal seek'`, I failing to reproduce the issue. Interestingly, it wasn't happening on the criu-dev branch.

**Comment by rst0git** [comment]
@nviennot Could you add [this commit](https://github.com/checkpoint-restore/criu/commit/60bf69bca50431ce213dcef2b4ec5e06c3245154) from the criu-dev branch in your PR?

File: criu/pipes.c (hunk 1 above)
Comment (nviennot) [reply criu/pipes.c:471]: Hello Mike, I don't think `-EAGAIN` is useful to check here as splice() is running in blocking mode here, as we don't pass `SPLICE_F_NONBLOCK`

File: criu/pipes.c [hunk 2]
```diff
@@ -463,18 +463,16 @@ int dump_one_pipe_data(struct pipe_data_dump *pd, int lfd, const struct fd_parms
...
-		} else if (wrote != bytes) {
-			pr_err("%#x: Wanted to write %d bytes, but wrote %d\n",
-					pipe_id(p), bytes, wrote);
-			goto err_close;
 		}
+		if (wrote == 0)
```
[the remark this answers is not in the document]
Comment (nviennot) [reply criu/pipes.c]: Yes

**Comment by nviennot** [comment]
> I'm curious how the --stream would interact with --lazy-pages?

I'm not sure. I don't see any reason the two couldn't work together, but it's not tested, so there might be bugs.

**Comment by nviennot** [comment]
@rst0git With the commit you suggested, the tests pass! Thank you :)

**Comment by avagin** [comment]
@nviennot could you check that I merged it properly: https://github.com/avagin/criu/tree/criu-streamer

**Comment by avagin** [comment]
Merged. Thanks a lot!

**Pull request closed without merging**